native/src/main/java/io/ballerina/stdlib/mcp/McpServiceMethodHelper.java

```diff
@@ -0,0 +1,195 @@
+     * @return           Result of remote method invocation.
+     */
+    public static Object invokeOnListTools(Environment env, BObject mcpService) {
+        return env.getRuntime().callMethod(mcpService, "onListTools", null);
+    }
+
+    /**
+     * Invoke the 'onCallTool' remote method on the given MCP service object with parameters.
+     *
+     * @param env        The Ballerina runtime environment.
+     * @param mcpService The MCP service object.
+     * @param params     Parameters for the tool invocation.
+     * @return           Result of remote method invocation.
+     */
+    public static Object invokeOnCallTool(Environment env, BObject mcpService, BMap<?, ?> params) {
+        return env.getRuntime().callMethod(mcpService, "onCallTool", null, params);
+    }
+
+    /**
+     * Lists tool metadata for remote functions in the given MCP service.
+     *
+     * @param mcpService The MCP service object.
+     * @param typed      The type descriptor for the result.
+     * @return           Record containing the list of tools.
+     */
+    public static Object listToolsForRemoteFunctions(BObject mcpService, BTypedesc typed) {
+        RecordType resultRecordType = (RecordType) typed.getDescribingType();
+        BMap<BString, Object> result = ValueCreator.createRecordValue(resultRecordType);
+
+        ArrayType toolsArrayType = (ArrayType) resultRecordType.getFields().get(FIELD_TOOLS).getFieldType();
+        BArray tools = ValueCreator.createArrayValue(toolsArrayType);
+
+        for (RemoteMethodType remoteMethod : getRemoteMethods(mcpService)) {
+            remoteMethod.getAnnotations().entrySet().stream()
+                    .filter(e -> e.getKey().getValue().contains(ANNOTATION_MCP_TOOL))
+                    .findFirst()
+                    .ifPresent(annotation -> tools.append(
+                            createToolRecord(toolsArrayType, remoteMethod, (BMap<?, ?>) annotation.getValue())
+                    ));
+        }
+        result.put(fromString(FIELD_TOOLS), tools);
+        return result;
+    }
+
+    /**
+     * Invokes a remote function (tool) by name with arguments.
+     *
+     * @param env        The Ballerina runtime environment.
+     * @param mcpService The MCP service object.
+     * @param params     The parameters for the tool invocation.
+     * @param typed      The type descriptor for the result.
+     * @return           Record containing the invocation result or an error.
+     */
+    public static Object callToolForRemoteFunctions(Environment env, BObject mcpService, BMap<?, ?> params,
+                                                    BTypedesc typed) {
+        BString toolName = (BString) params.get(fromString(FIELD_NAME));
+
+        RemoteMethodType method = getRemoteMethods(mcpService).stream()
+                .filter(rmt -> rmt.getName().equals(toolName.getValue()))
+                .findFirst().orElse(null);
+
+        if (method == null) {
```

Comment:
Rather than performing null checks, consider using Optional and checking for isEmpty(), similar to the approach used in the createCallToolResult method.